lib/git_worktree.rb

```diff
@@ -0,0 +1,327 @@
+require 'rugged'
+
+class GitWorktree
+  attr_accessor :name, :email, :base_name
+  ENTRY_KEYS = [:path, :dev, :ino, :mode, :gid, :uid, :ctime, :mtime]
+  DEFAULT_FILE_MODE = 0100644
+  LOCK_REFERENCE = 'refs/locks'
+
+  def initialize(options = {})
+    raise ArgumentError, "Must specify path" unless options.key?(:path)
+    @path          = options[:path]
+    @email         = options[:email]
+    @name          = options[:name]
+    @bare          = options[:bare]
+    @commit_sha    = options[:commit_sha]
+    @password      = options[:password]
+    @fast_forward_merge = options[:ff] || true
+    @ssl_no_verify = options[:ssl_no_verify] || false
+    @remote_name   = 'origin'
+    @cred          = Rugged::Credentials::UserPassword.new(:username => @name,
+                                                           :password => @password)
+    @base_name     = File.basename(@path)
+
+    process_repo(options)
+  end
+
+  def delete_repo
+    return false unless @repo
+    @repo.close
+    FileUtils.rm_rf(@path)
+    true
+  end
+
+  def add(path, data, commit_sha = nil, default_entry_keys = {})
+    entry = {}
+    entry[:path] = path
+    ENTRY_KEYS.each { |key| entry[key] = default_entry_keys[key] if default_entry_keys.key?(key) }
+    entry[:oid]  = @repo.write(data, :blob)
+    entry[:mode] ||= DEFAULT_FILE_MODE
+    entry[:mtime] ||= Time.now
+    current_index(commit_sha).add(entry)
+  end
+
+  def remove(path, commit_sha = nil)
+    current_index(commit_sha).remove(path)
+  end
+
+  def remove_dir(path, commit_sha = nil)
+    current_index(commit_sha).remove_dir(path)
+  end
+
+  def file_exists?(path, commit_sha = nil)
+    find_entry(path, commit_sha) ? true : false
+  end
+
+  def directory_exists?(path, commit_sha = nil)
+    entry = find_entry(path, commit_sha)
+    entry && entry[:type] == :tree
+  end
+
+  def read_file(path, commit_sha = nil)
+    read_entry(fetch_entry(path, commit_sha))
+  end
+
+  def read_entry(entry)
+    @repo.lookup(entry[:oid]).content
+  end
+
+  def directory?(path, commit_sha = nil)
+    entry = fetch_entry(path, commit_sha)
+    entry[:type] == :tree
+  end
+
+  def entries(path, commit_sha = nil)
+    tree = path.empty? ? lookup_commit_tree(commit_sha || @commit_sha) : get_tree(path, commit_sha)
+    tree.find_all.collect { |e| e[:name] }
+  end
+
+  def nodes(path, commit_sha = nil)
+    tree = path.empty? ? lookup_commit_tree(commit_sha || @commit_sha) : get_tree(path, commit_sha)
+    entries = tree.find_all
+    entries.each do |entry|
+      entry[:full_name] = File.join(@base_name, path, entry[:name])
+      entry[:rel_path] = File.join(path, entry[:name])
+    end
+  end
+
+  def save_changes(message, owner = :local)
+    cid = commit(message)
+    owner == :local ? lock { merge(cid) } : merge_and_push(cid)
+    true
+  end
+
+  def commit(message)
+    tree = @current_index.write_tree(@repo)
+    parents = @repo.empty? ? [] : [@repo.ref('refs/heads/master').target].compact
```

Comment:
`@repo.ref('refs/heads/master').target` sounds like it's worth extracting (or maybe with `.try(:target)`) as a private method